[Dependency Cache] Enable Dependency Cache on CI

ParaskP7 opened the pull request:
<!-- Remember about a good descriptive title. -->

Project Thread: paaHJt-7CE-p2
Related: #13170 + #13122 + #13288
Depends On: [a8c-ci-toolkit#135](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/135) -> [a8c-ci-toolkit#138](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/138)
Required By: [buildkite-ci#576](https://github.com/Automattic/buildkite-ci/pull/576)

<!-- Id number of the GitHub issue this PR addresses. -->

---

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR enables dependency cache on CI for this project. 💾

The dependency cache mechanism is split between:
- A weekly scheduled **save cache** build (see [commit](https://github.com/woocommerce/woocommerce-android/pull/13303/commits/684c6e6d8d1a8b3942705a10e4a43c3b2f7a484c) and [buildkite-ci#576](https://github.com/Automattic/buildkite-ci/pull/576) PR), which targets the below tasks only:
  - Mobile App (`WooCommerce:assembleJalapenoDebug`)
  - Wear App (`WooCommerce-Wear:assembleJalapenoDebug`)
  - Lint (`WooCommerce:lintJalapenoDebug`)
  - Unit Tests (`ssembleJalapenoDebugUnitTest assembleDebugUnitTest`)
  - Android Tests (`assembleJalapenoDebugAndroidTest`)
- A **restore cache** step that is run on targeted [pipeline.yml](https://github.com/woocommerce/woocommerce-android/blob/trunk/.buildkite/pipeline.yml) related jobs, via their corresponding individual scripts:
  - Mobile App (`prototype-build.sh`)
  - Wear App (`prototype-build.sh`)
  - Lint (`lint.sh`)
  - Unit Tests (`run-unit-tests.sh`)
  - Android tests (`run-instrumented-tests.sh`)

PS.1: The `detekt` job isn't targeted because it only takes about 20+ seconds of network request to download all the dependencies needed for this specific job ([build](https://buildkite.com/automattic/woocommerce-android/builds/26561#01946578-e466-4438-be6a-0a67488f8395) + [scan](https://scans.gradle.com/s/ftu4zklrffa3q/performance/network-activity)), which is about the same time it takes for the dependency cache to actually be restored, which is more or less about 10+ seconds ([build](https://buildkite.com/automattic/woocommerce-android/builds/26609#019469e4-c246-4c75-9808-31925b53689b/189-198)). Thus, this optimization is not really worth it for this specific job. For all those other jobs targeted, it takes more than a minute of network requests to download all the dependencies needed ([build](https://buildkite.com/automattic/woocommerce-android/builds/26561#01946578-e46c-4e6b-9157-596bd033d26f) + [scan](https://scans.gradle.com/s/fbw7fyc4enqw2/performance/network-activity)), and as such the diff is worth it.

PS.2: Dependency cache isn't enabled on the [release-builds.yml](https://github.com/woocommerce/woocommerce-android/blob/trunk/.buildkite/release-builds.yml) pipeline just yet, but just temporarily, and until we become comfortable with running this new process on the main [pipeline.yml](https://github.com/woocommerce/woocommerce-android/blob/trunk/.buildkite/pipeline.yml) pipeline.

---

### Future Improvements

As per [this](https://github.com/woocommerce/woocommerce-android/pull/13303#discussion_r1916744733) comment, me and @wzieba has been discussing on DMs about some possible future improvement, which are including but are not limited to the below list:

1. Only execute the `assemble android test` task, which depends on `assemble` task anyway, and would most likely resolve most of the dependencies, for all other tasks to benefit from. However, this needs further testing to make sure that running other tasks such us `lint` and `assemble unit test` are providing little to no help.
2. Create a custom Gradle task and utilize Gradle's [init scripts](https://docs.gradle.org/current/userguide/init_scripts.html) (`init.d`). Using the init script, go through all available modules for a project, and then, using that custom Gradle task, resolve all dependencies for all tasks that are available per module. We could take some inspiration from plugins such as [this](https://plugins.gradle.org/plugin/org.gradle.github-dependency-graph-gradle-plugin).
3. Instead of scheduling the `dependency cache` build to run weekly ([buildkite-ci#576](https://github.com/Automattic/buildkite-ci/pull/576)), make it run more often, possibly scheduling it to even run daily. However, since dependencies are not being updated too often, and, even when they do on that specific week, it would only be for a few of them. This will indeed cause some extra network requests, but most of the dependencies cached would still benefit from the existing and weekly refreshed dependency cache.
4. Enable dependency cache on the [release-builds.yml](https://github.com/woocommerce/woocommerce-android/blob/trunk/.buildkite/release-builds.yml) pipeline.

---

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

This dependency cache solution was thoroughly tested via this #13170 test-only PR, but to help you asserting that this change is indeed making an impact, plus not causing any trouble for the project, try and follow the below:

- `Save Cache`:

  - Triggering [schedules/dependency-cache.yml](https://github.com/woocommerce/woocommerce-android/pull/13303/files#diff-f1eb78f1f1134bf64e6d4a27fd0b816ae7dff3a0da45f4115e8898bc6939642e) manually via [this](https://buildkite.com/automattic/woocommerce-android/builds/26606) build:
    - Starts [downloading the dependencies per task](https://buildkite.com/automattic/woocommerce-android/builds/26606#019469da-aebc-4dc3-a4d8-e3b967c76b41/187).
    - Caches [the dependencies successfully](https://buildkite.com/automattic/woocommerce-android/builds/26606#019469da-aebc-4dc3-a4d8-e3b967c76b41/9110-9111).
      - Duration: [42s](https://buildkite.com/automattic/woocommerce-android/builds/26606#019469da-aebc-4dc3-a4d8-e3b967c76b41/9110-9121)
      - Size: [625.74 MB](https://buildkite.com/automattic/woocommerce-android/builds/26606#019469da-aebc-4dc3-a4d8-e3b967c76b41/9110-9122)
    - Total time it takes for the `dependency cache` job to complete: [7m 42s](https://buildkite.com/automattic/woocommerce-android/builds/26606#019469da-aebc-4dc3-a4d8-e3b967c76b41)
  - Sending a Slack notification (`#android-core-notifs`):
    - On success, notifying that: `The dependency cache has been updated.` (p1736943241087289-slack-C0787KRDYDC)
    - On failure, notifying that: `Failure to update the dependency cache.` (not tested, but should behave similarly to ☝️)
- `Restore Cache`:

| Job | BT (Without) | BT (With) | BT (Diff*) | NR (Without) | NR (With) | RC (Without) | RC (With) | Win |
| ---- | ------------ | ---------- | ---- | ------------- | --------- | -------------- | --------- | ---- |
| Mobile App | [5m 38s](https://buildkite.com/automattic/woocommerce-android/builds/26561#01946578-e46c-4e6b-9157-596bd033d26f) | [3m 53s](https://buildkite.com/automattic/woocommerce-android/builds/26609#019469e4-c246-4c75-9808-31925b53689b) | `- 1m 45s` (✅) | [1m 15s](https://scans.gradle.com/s/fbw7fyc4enqw2/performance/network-activity) | [0s](https://scans.gradle.com/s/xltgddi566p4g/performance/network-activity) | `0s` | [12s](https://buildkite.com/automattic/woocommerce-android/builds/26609#019469e4-c246-4c75-9808-31925b53689b/189-198) | `1m 3s` (✅) |
| Wear App | [4m 43s](https://buildkite.com/automattic/woocommerce-android/builds/26561#01946578-e46d-4415-8e6f-0a2f6245333b) | [2m 56s](https://buildkite.com/automattic/woocommerce-android/builds/26609#019469e4-c247-4869-b846-4a69ac9675c1) | `- 1m 47s` (✅) | [59s](https://scans.gradle.com/s/6hpqcu2spugwg/performance/network-activity) | [0s](https://scans.gradle.com/s/tfyjvmh33szxe/performance/network-activity) | `0s` | [15s](https://buildkite.com/automattic/woocommerce-android/builds/26609#019469e4-c247-4869-b846-4a69ac9675c1/189-198) | `44s` (✅) |
| Lint | [6m 9s](https://buildkite.com/automattic/woocommerce-android/builds/26561#01946578-e469-4f0b-8c79-0bbf0aef92c1) | [4m 6s](https://buildkite.com/automattic/woocommerce-android/builds/26609#019469e4-c242-4629-8651-6b240d97f4de) | `- 2m 3s` (✅) | [1m 49s](https://scans.gradle.com/s/3vodnbxajkqug/performance/network-activity) | [0s](https://scans.gradle.com/s/nstv7ejuxme4y/performance/network-activity) | `0s` | [14s](https://buildkite.com/automattic/woocommerce-android/builds/26609#019469e4-c242-4629-8651-6b240d97f4de/189-198) | `1m 35s` (✅) |
| Unit Tests | [6m 0s](https://buildkite.com/automattic/woocommerce-android/builds/26561#01946578-e470-4846-b3bf-3c65b0f3e6c3) | [3m 57s](https://buildkite.com/automattic/woocommerce-android/builds/26609#019469e4-c24a-4bb8-9a3e-24052f20e33c) | `- 2m 3s` (✅) | [1m 31s](https://scans.gradle.com/s/uusmmn4gmwymw/performance/network-activity) | [0s](https://scans.gradle.com/s/5u33enw43hjec/performance/network-activity) (ish) | `0s` | [10s](https://buildkite.com/automattic/woocommerce-android/builds/26609#019469e4-c24a-4bb8-9a3e-24052f20e33c/224-233) | `1m 21s` (✅) |
| UI Tests | [10m 3s](https://buildkite.com/automattic/woocommerce-android/builds/26561#01946578-e471-4783-abf1-c5f08f04612f) | [13m 37s](https://buildkite.com/automattic/woocommerce-android/builds/26609#019469e4-c24a-454f-a98f-f81c82f2c908) | `+ 3m 34s` (❗) | [1m 38s](https://scans.gradle.com/s/stvpooxfp3bnm/performance/network-activity) | [0s](https://scans.gradle.com/s/eaedij3zqpbio/performance/network-activity) | `0s` | [15s](https://buildkite.com/automattic/woocommerce-android/builds/26609#019469e4-c24a-454f-a98f-f81c82f2c908/224-233) | `1m 23s` (✅) |

> [!NOTE]
> - `Without`: Without Dependency Cache
> - `With`: With Dependency Cache
> - `BT`: Build Time
> - `NR`: Network Requests
> - `RC`: Restore Cache
> - `Win`: NR (Without) - RC (With)

> [!WARNING]
> (*): The `BT (Diff)` presented above is relative, so don't read too much into it, it might not match with the `WIN`, which is a much more solid number. But still, it is nice to have and compare those two number together. To to be clear about that, although I tried to pick the most representative builds, without and with the dependency cache mechanism in place, there are other factors that can make any build faster, or slower for that matter (like `build cache`, `firebase test lab`, etc), so again, don't read too much into it.
 
---

### Merge Instructions

- [x] Wait for [a8c-ci-toolkit#135](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/135) -> [a8c-ci-toolkit#138](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/138) to be reviewed, approved and merged.
- [x] Create a new [a8c-ci-toolkit release](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/releases) ([3.9.0](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/releases/tag/3.9.0) -> [3.9.1](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/releases/tag/3.9.1) ).
- [x] Update [CI_TOOLKIT](https://github.com/woocommerce/woocommerce-android/blob/trunk/.buildkite/shared-pipeline-vars#L6) to point to this newly created [a8c-ci-toolkit release](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/releases) (3547e664e43bcb37d7e5c67190aa68bf7f492842 -> 6ac4eb4555927d660994fb8bb7e2fc3aae6faad6).
- [x] Remove `status: do not merge` label.
- [x] Merge this PR.
- [x] Close #13122 and #13170 test-only PRs.

---

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

ParaskP7 committed:
CI: Update a8c-ci-toolkit to pr-sha commit (dependency cache)

FYI: This change is done for testing purposes and until the below
'A8C CI Toolkit' #135 PR gets merged to 'trunk'. When that's
done, the 'a8c-ci-toolkit' will be updated to '3.9.0' instead.

A8C CI Toolkit PR: [Dependency Updates] [Dependency Cache]
Dependency Cache on CI per Project [without GRADLE_RO_DEP_CACHE] #135
- https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/135

ParaskP7 committed:
CI: Add a scheduled dependency cache job (save cache)

FYI: This job will be then used by 'buildkite-ci' and configured as a
'buildkite_pipeline_schedule' with a weekly frequency.

PS: The targeted 'pipeline' related jobs are:
- Mobile App
- Wear App
- Lint
- Unit Tests
- Android tests

ParaskP7 committed:
CI: Remove unnecessary echo linting from detekt pipeline job

ParaskP7 committed:
CI: Add restore cache and run on targeted pipeline related jobs only

PS: The targeted 'pipeline' related jobs are:
- Mobile App
- Wear App
- Lint
- Unit Tests
- Android tests

PS: The 'detekt' job isn't targeted because it only takes about 20+
seconds of network request to download all the dependencies needed for
this specific job, which is about the same time it take for the
dependency cache to actually be restored (10+ seconds). Thus, this
optimization is not really worth it for this specific job. For all those
other jobs targeted, it take more than a minute of network requests to
download all the dependencies needed, and as such worth the diff is
worth it.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ParaskP7 committed:
CI: Update a8c-ci-toolkit to 3.9.0

Release Notes: https://github.com/Automattic/
a8c-ci-toolkit-buildkite-plugin/releases/tag/3.9.0

Related: 4475a6ad4bb6aa6be778c4b37374d87f92f717d2

ParaskP7 committed:
CI: Update a8c-ci-toolkit to pr-sha commit (fix dependency cache)

FYI: This change is done for testing purposes and until the below
'A8C CI Toolkit' #138 PR gets merged to 'trunk'. When that's
done, the 'a8c-ci-toolkit' will be updated to '3.9.1' instead.

A8C CI Toolkit PR: [Dependency Cache] Fix Dependency Cache #138
- https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/138

ParaskP7 committed:
CI: Update a8c-ci-toolkit to 3.9.1

Release Notes: https://github.com/Automattic/
a8c-ci-toolkit-buildkite-plugin/releases/tag/3.9.1

Related: 49d771a9cf4f11e7efdf7dd0641e2b529bf39f07

wzieba commented on .buildkite/commands/save-cache.sh:
```diff
@@ -0,0 +1,32 @@
+echo ""
+
+echo "--- 🧪 Download Android Test Dependencies [Assemble Android Tests]"
+./gradlew assembleJalapenoDebugAndroidTest
```
As discussed on DMs, I think we could have only this task executed (plus maybe the one for unit tests) but we also agreed it's okay to have this like this for now, and maybe think about even a custom Gradle task for this 👍

ParaskP7 replied:
Oh yea @wzieba , and actually thanks for reminding me about it cause I forgot to add a `Potential Future Improvements` section on the PR description itself, with all what we've discussed on DMs. 🙇

ParaskP7 replied:
@wzieba I just added the below section on the PR description, let me know if I missed something important: 🙏 

![image](https://github.com/user-attachments/assets/84527d45-f1e9-472e-8dfa-8e208f1a365d)

ParaskP7 replied:
FYI: Plus this (4.) improvement I just added on the PR description, just for completeness purposes and so we don't forget about it:

![image](https://github.com/user-attachments/assets/87f2d168-6180-4bbe-82fb-f1375d813847)

wzieba commented on .buildkite/commands/save-cache.sh:
```diff
@@ -0,0 +1,32 @@
+echo ""
+
+echo "--- 🧪 Download Unit Test Dependencies [Assemble Unit Tests]"
+./gradlew assembleJalapenoDebugUnitTest lib:cardreader:assembleDebugUnitTest
```
This will not cache dependencies for all submodules, I think we want rather this 
```suggestion
./gradlew assembleJalapenoDebugUnitTest assembleDebugUnitTest
```

ParaskP7 replied:
Oh, thanks for noticing that @wzieba , I was trying to have this on a 1-on-1 relationship with this corresponding [run-unit-tests.sh](https://github.com/woocommerce/woocommerce-android/blob/trunk/.buildkite/commands/run-unit-tests.sh) script, but it was recently changed (758606435bdb7287b8c680a46a005ab15f723823) as part of this #13052 PR into exactly [that](https://github.com/woocommerce/woocommerce-android/blob/trunk/.buildkite/commands/run-unit-tests.sh#L5), I'll proceed with the change. 💯 

PS: Btw, for some reason I was thinking, and I could actually swear that I tested it in the past, that running `assembleDebugUnitTest` will also trigger `assembleJalapenoDebugUnitTest`, plus `assembleWasabiDebugUnitTest` and `assembleVanillaDebugUnitTest`, because of all the different variants, but actually, me testing it locally again, it doesn't looks like it. 👍

ParaskP7 replied:
Done: cbca6ac19fcc9320bccaac54d2a4dcd227a4fbb1 + [build](https://buildkite.com/automattic/woocommerce-android/builds/26616)

wzieba commented on .buildkite/pipeline.yml:
```diff
@@ -35,15 +35,15 @@ steps:
       - label: "lint"
-        command: .buildkite/commands/lint.sh
+        command: |
+          .buildkite/commands/restore-cache.sh
+          .buildkite/commands/lint.sh
```
I think it might be actually better to move `restore-cache` call to `lint.sh` (and in other cases as well).

These scripts (`lint.sh` and the rest) intentionally extracted all logic regarding job execution, so I don't think we should leak some of the logic to `pipeline.yml`. It's not a big deal though.

ParaskP7 replied:
Hmmm, maybe you're right @wzieba ... 🤔 

I was thinking that this way we are being explicit and using `restore-cache.sh` exactly where needed, without hiding it within a specific script like `lint.sh`, where it could be easily missed. But, this way we are creating more boilerplate and adding more logic into `pipeline.yml`.

On the other hand, if we move `restore-cache.sh` into a specific script like `lint.sh`, we are avoiding adding more boilerplate into `pipeline.yml`, reducing any possible duplication (ie. `prototype-build.sh` is call twice) and having `restore-cache.sh` run on any pipeline/job that might be using that specific script (maybe unintentional). The only downside is that we are now hiding `restore-cache.sh` deeper, making it less explicit, and maybe combining two different processes into one script, which we better kept those separate.

I am still a bit torn on which of those approaches is better... 😊

ParaskP7 replied:
Done: 4798b6346c703359fd7b5e1aa929d537b379139e

FYI: After thinking a bit more on this topic, I actually decided to take your advice. 👍

PS: In order to connect the dots a bit more, just to have this be a bit more explicit, [something I worried about](https://github.com/woocommerce/woocommerce-android/pull/13303#discussion_r1916833555), I also added extra comments on top of each Gradle task run within the `save-cache.sh` script. This is so that readers would know how each `save` script/task(s) is connected to its corresponding `restore` script/task(s). I hope you'll like this change, let me know what you think @wzieba . 😊

wzieba replied:
Sure, looks good 👍

ParaskP7 replied:
Awesome, thanks for the confirmation on that @wzieba , this is now ready to merge! 🚀

ParaskP7 committed:
CI: Update save cache to run assemble debug unit test on all lib modules

Co-authored-by: Wojciech Zięba <wojtek.zieba@automattic.com>

ParaskP7 committed:
CI: Move restore cache call to individual .sh script files

PR Comment: https://github.com/woocommerce/woocommerce-android/pull/
13303#discussion_r1916752265

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into ci/enable-dependency-cache-on-ci-final

ParaskP7 committed:
CI: Fix wrongly defined command related double quotes on pipeline jobs

When using 'single-cmd' and not using this double quotes structure it is
is producing 'did not find expected key' type of errors.

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into ci/enable-dependency-cache-on-ci-final
